src/mkdocstrings/handlers/python.py

```diff
@@ -290,3 +295,33 @@ def rebuild_category_lists(obj: dict) -> None:
     obj["children"] = [child for _, child in obj["children"].items()]
     for child in obj["children"]:
         rebuild_category_lists(child)
+
+
+def sort_object(obj: CollectorItem, sort_style: str) -> None:
+    """Sort the collected object's children.
+
+    Sorts the object's children list, then each category separately, and then recurses into each.
+
+    Arguments:
+        obj: The collected object, as a dict. Note that this argument is mutated.
+        sort_style: How to sort the children lists - 'alphabetical' or 'source'.
+    """
+
+    if sort_style == "alphabetical":
+        # sort_last is a string that contains the final unicode character, so
+        # if 'name' isn't found on the object, the item will go to the end of
+        # the list.
+        sort_last = chr(sys.maxunicode)
+        sort_function = lambda item: item.get("name", sort_last)
+    elif sort_style == "source":
+        sort_function = lambda item: item.get("source", {}).get("line_start", 0)
+    else:
+        raise PluginError('unknown sort_style')
```

Comment:
@oprypin @joerick do you think we should optimize this again by refactoring the two lambdas into proper (private) functions? One function or the other would then be passed directly to `sort_object` instead of re-computing it in each step of the recursion. I know it might be premature optimization, but it's easy to do, so why not.

This would also allow getting the PluginError out of this function.